comm.go

```diff
@@ -88,22 +87,29 @@ func (p *PubSub) handleNewStream(s network.Stream) {
 	}
 }
 
+func (p *PubSub) notifyPeerDead(pid peer.ID) {
+	p.peerDeadPrioLk.RLock()
+	p.peerDeadMx.Lock()
+	p.peerDeadPend[pid] = struct{}{}
+	p.peerDeadMx.Unlock()
+	p.peerDeadPrioLk.RUnlock()
+
+	select {
+	case p.peerDead <- struct{}{}:
+	default:
+	}
+}
+
 func (p *PubSub) handleNewPeer(ctx context.Context, pid peer.ID, outgoing <-chan *RPC) {
 	s, err := p.host.NewStream(p.ctx, pid, p.rt.Protocols()...)
 	if err != nil {
 		log.Debug("opening new stream to peer: ", err, pid)
 
-		var ch chan peer.ID
-		if err == ms.ErrNotSupported {
-			ch = p.newPeerError
-		} else {
-			ch = p.peerDead
```

Comment:
Why are we getting rid of this? Can't we fail to create a new stream if, e.g., the connection is dead? This will leave us without a peer if we try to use the wrong connection.

Reply:
it can lead to some very nasty "peer declared dead; respawning writers" loops, which have been reported by lotus users in the wild.

Note sure how to properly handle this situation, let me give it a bit more thought -- it's subtle and what we have right now is very problematic.

Reply:
and that's why I removed it.

Reply:
Ah, I see. Yeah. This is a crap situation. I'm just concerned that this will lead to the _opposite_ problem, but I think this is fine for now. But we should put in a better fix later.

Reply:
Yeah, I am concerned about it too. I'll open a follow up issue so that we don't forget.

Reply:
follow up issue in https://github.com/libp2p/go-libp2p-pubsub/issues/433